geoutils/raster/raster.py

```diff
@@ -3363,96 +3443,186 @@ def split_bands(self: RasterType, copy: bool = False, bands: list[int] | int | N
-        if subsample <= 0.0:
-            raise ValueError(f"subsample cannot be zero or negative (given value: {subsample})")
-        # If the sample is equal to or less than 1, it is assumed to be a fraction.
-        if subsample <= 1.0:
-            sample = int(data_size * subsample)
-        else:
-            sample = int(subsample)
-        if sample > data_size:
-            raise ValueError(f"sample cannot exceed the size of the dataset ({sample} vs {data_size})")
+        # Input checks
+        if not isinstance(data_column_name, str):
+            raise ValueError("Data column name must be a string.")
+        if not isinstance(data_band, int) and (1 < data_band or self.count < data_band):
+            raise ValueError(f"Data band number must be between 1 and the total number of bands of {self.count}.")
+
+        # Rename data column if a different band is selected but the name is still default
+        if data_band != 1 and data_column_name == "b1":
+            data_column_name = "b" + str(data_band)
 
-        # If the sample is smaller than the max size, take a random sample of indices, otherwise take the whole.
-        choice = np.random.randint(0, data_size - 1, sample) if sample != data_size else np.arange(data_size)
+        # We do 2D subsampling on the data band only, regardless of valid masks on other bands
+        if self.is_loaded:
+            if self.count == 1:
+                self_mask = get_mask_from_array(self.data)  # This is to avoid the case where the mask is just "False"
+            else:
+                self_mask = get_mask_from_array(
+                    self.data[data_band - 1, :, :]
+                )  # This is to avoid the case where the mask is just "False"
+            valid_mask = ~self_mask
 
-        cols = choice % self.width
-        rows = (choice / self.width).astype(int)
+        # Load only mask of valid data from disk if array not loaded
+        else:
+            valid_mask = ~self._load_only_mask(bands=data_band)
+
+        # Get subsample on valid mask
+        # Build a low memory boolean masked array with invalid values masked to pass to subsampling
+        ma_valid = np.ma.masked_array(data=np.ones(np.shape(valid_mask), dtype=bool), mask=~valid_mask)
+        # Take a subsample within the valid values
+        indices = subsample_array(array=ma_valid, subsample=subsample, random_state=random_state, return_indices=True)
 
-        # Extract the coordinates of the pixels and filter by the chosen pixels.
-        x_coords, y_coords = (np.array(a) for a in self.ij2xy(rows, cols, offset=pixel_offset))
+        # Extract the coordinates at subsampled pixels with valid data
+        # To extract data, we always use "upper left" which rasterio interprets as the exact coordinates of the raster
+        # Further below we redefine output coordinates based on point interpretation
+        x_coords, y_coords = (np.array(a) for a in self.ij2xy(indices[0], indices[1], offset="ul"))
```

Comment:
It seems like `x_coords` and `y_coords` are only used in the second block of the following if statement, so could be moved there no?

Reply:
Yes, done!